Fix dynamo issue "Failed running call_function <built-in method sparse_coo_tensor of type object at 0xDEADBEEF"

aorenste opened the pull request:
A small medley of fixes:
- When validating sparse tensor indices don't check numel() if it's symbolic.
- When validating sparse tensor indices if the indices are a FakeTensor then don't require them to actually exist.
- Added TensorVariable.method_coalesce() implementation (unimplemented so it graph breaks - correct since the output tensor is dependent on the input data)
- merge_view_inputs() shouldn't be looking at storage of sparse tensors
- In insertable_tensor_check() don't treat sparse tensors as uniform.

And some drive-bys:
- Spelling nits
- Moved error text out of header file
- In has_metadata_mutation() only check StorageWeakRef if the result is going to be used.

Fixes #119780

This fixes about 71 of the 110 errors in #119780. The remaining issues hit follow-on errors (and should probably get new tasks once we refresh the error list)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118192
* #120206



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

aorenste committed:
WIP: initial support of test/test_testing.py -v -k test_matching_coalesced

[ghstack-poisoned]

aorenste committed:
Update on "WIP: initial support of test/test_testing.py -v -k test_matching_coalesced"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aorenste committed:
Update on "WIP: initial support of test/test_testing.py -v -k test_matching_coalesced"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aorenste committed:
Update on "WIP: initial support of test/test_testing.py -v -k test_matching_coalesced"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aorenste committed:
Update on "WIP: initial support of test/test_testing.py -v -k test_matching_coalesced"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aorenste committed:
Update on "WIP: initial support of test/test_testing.py -v -k test_matching_coalesced"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aorenste committed:
Update on "WIP: initial support of test/test_testing.py -v -k test_matching_coalesced"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aorenste committed:
Update on "WIP: initial support of test/test_testing.py -v -k test_matching_coalesced"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aorenste committed:
Update on "WIP: initial support of test/test_testing.py -v -k test_matching_coalesced"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

aorenste committed:
Update on "Fix validation of sparse tensors when indices are a FakeTensor."

When validating sparse tensor indices if the indices are a FakeTensor then don't require them to actually exist.

We actually create the indices as a real tensor but then it gets lifted to a FakeTensor - which doesn't have data.

Fixes #119780

But introduces #120413
#progress?




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

aorenste committed:
Update on "Fix dynamo issue "Failed running call_function <built-in method sparse_coo_tensor of type object at 0xDEADBEEF""

A small medley of fixes:
- When validating sparse tensor indices don't check numel() if it's symbolic.
- When validating sparse tensor indices if the indices are a FakeTensor then don't require them to actually exist.
- Added TensorVariable.method_coalesce() implementation (unimplemented so it graph breaks - correct since the output tensor is dependent on the input data)
- merge_view_inputs() shouldn't be looking at storage of sparse tensors
- In insertable_tensor_check() don't treat sparse tensors as uniform.

And some drive-bys:
- Spelling nits
- Moved error text out of header file
- In has_metadata_mutation() only check StorageWeakRef if the result is going to be used.

Fixes #119780

This fixes about 71 of the 110 errors in #119780. The remaining issues hit follow-on errors (and should probably get new tasks once we refresh the error list)




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

zou3519 commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -867,7 +867,7 @@ def _same_dtype_views(view1, view2):
     other_args = []
     for i, inpt in enumerate(fwd_inputs):
-        if isinstance(inpt, Tensor):
+        if isinstance(inpt, Tensor) and not inpt.is_sparse:
```
meta point:
- This PR looks like it is adding support for sparse tensors to torch.compile
- "adding sparse tensor support to torch.compile" likely needs a design doc and discussion because of the amount of complexity. the PT2 stack assumes it is working on regular strided Tensors (except for the case of torch_dispatch / torch_function Tensor subclasses, which Sparse tensors are not)
- One "easy" way to fix the tests is to just have Dynamo graph break if it sees a sparse tensor

aorenste replied:
Not really adding support - just making some things that previously failed hard either trivially pass or fail less hard.

ezyang replied:
This particular change seems problematic though. It probably would be better for AOTAutograd to raise an error if it receives a sparse tensor input. Not sure how this affects how many tests start passing though.

pearu replied:
FWIW, the `.is_sparse` check covers only sparse COO tensors but we have also sparse compressed tensors. Here, I suggest using `inpt.layout is torch.strided` instead of `inpt.is_sparse`.

aorenste committed:
Update on "Fix dynamo issue "Failed running call_function <built-in method sparse_coo_tensor of type object at 0xDEADBEEF""

A small medley of fixes:
- When validating sparse tensor indices don't check numel() if it's symbolic.
- When validating sparse tensor indices if the indices are a FakeTensor then don't require them to actually exist.
- Added TensorVariable.method_coalesce() implementation (unimplemented so it graph breaks - correct since the output tensor is dependent on the input data)
- merge_view_inputs() shouldn't be looking at storage of sparse tensors
- In insertable_tensor_check() don't treat sparse tensors as uniform.

And some drive-bys:
- Spelling nits
- Moved error text out of header file
- In has_metadata_mutation() only check StorageWeakRef if the result is going to be used.

Fixes #119780

This fixes about 71 of the 110 errors in #119780. The remaining issues hit follow-on errors (and should probably get new tasks once we refresh the error list)




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

aorenste committed:
Update on "Fix dynamo issue "Failed running call_function <built-in method sparse_coo_tensor of type object at 0xDEADBEEF""

A small medley of fixes:
- When validating sparse tensor indices don't check numel() if it's symbolic.
- When validating sparse tensor indices if the indices are a FakeTensor then don't require them to actually exist.
- Added TensorVariable.method_coalesce() implementation (unimplemented so it graph breaks - correct since the output tensor is dependent on the input data)
- merge_view_inputs() shouldn't be looking at storage of sparse tensors
- In insertable_tensor_check() don't treat sparse tensors as uniform.

And some drive-bys:
- Spelling nits
- Moved error text out of header file
- In has_metadata_mutation() only check StorageWeakRef if the result is going to be used.

Fixes #119780

This fixes about 71 of the 110 errors in #119780. The remaining issues hit follow-on errors (and should probably get new tasks once we refresh the error list)




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

ezyang commented:
Related https://github.com/pytorch/pytorch/pull/120498 cc @pearu

ezyang commented on torch/_functorch/_aot_autograd/functional_utils.py:
```diff
@@ -162,10 +162,13 @@ def has_metadata_mutation(f_arg, arg, *, check_only_storage_mutation: bool):
-            arg_after.untyped_storage()
-        )
-        has_storage_metadata_mutation = maybe_storage_changed and not same_storages
+        if maybe_storage_changed:
```
@bdhirsh please review

ezyang commented:
This all seems basically reasonable.

pearu commented on aten/src/ATen/native/sparse/SparseTensor.cpp:
```diff
@@ -379,41 +379,46 @@ void _validate_sparse_coo_tensor_args(
+    // Some meta tensors (especially FakeTensor) may not have data available. We can't
+    // use `isTensorSubclassLike` because some tensor subclasses DO have data and we have
+    // tests that expect these checks to run (test_factory_size_check w/ crossref).
+    if (indices.has_storage() && indices.storage().data()) {
```
@ezyang , in short, no. The PR https://github.com/pytorch/pytorch/pull/120498 is about CSR/CSC/BSR/BSC meta tensors but here we are considering COO fake tensors. That said, the approach used in #120498 applies also here: the validation of sparse meta/fake tensors should skip invariant checks that access indices, however, invariant checks about shapes/devices/dtypes still make sense for meta/fake tensors.

In the case of meta tensors, it is sufficient to use `if (indices.is_meta())`, but for fake tensors we'll need a more general method, for instance:
```c++
if (indices.storage().device_type() == kMeta) {
  ...
}
```
that should apply to both meta and fake tensors (untested).

Considering https://github.com/pytorch/pytorch/blob/834c7a1d3ea07878ad87d127ee28606fc140b552/test/test_fake_tensor.py#L1064 that implementation is wrong in many aspects, as a next step, I planned to fix this test and then implement fake tensors support for sparse compressed tensors as well as address any issues with COO fake tensors.

ezyang replied:
okey. I wonder if we suppress this check for tensor subclasses in general, that would have worked for fake too

aorenste committed:
Update on "Fix dynamo issue "Failed running call_function <built-in method sparse_coo_tensor of type object at 0xDEADBEEF""

A small medley of fixes:
- When validating sparse tensor indices don't check numel() if it's symbolic.
- When validating sparse tensor indices if the indices are a FakeTensor then don't require them to actually exist.
- Added TensorVariable.method_coalesce() implementation (unimplemented so it graph breaks - correct since the output tensor is dependent on the input data)
- merge_view_inputs() shouldn't be looking at storage of sparse tensors
- In insertable_tensor_check() don't treat sparse tensors as uniform.

And some drive-bys:
- Spelling nits
- Moved error text out of header file
- In has_metadata_mutation() only check StorageWeakRef if the result is going to be used.

Fixes #119780

This fixes about 71 of the 110 errors in #119780. The remaining issues hit follow-on errors (and should probably get new tasks once we refresh the error list)




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

aorenste committed:
Update on "Fix dynamo issue "Failed running call_function <built-in method sparse_coo_tensor of type object at 0xDEADBEEF""

A small medley of fixes:
- When validating sparse tensor indices don't check numel() if it's symbolic.
- When validating sparse tensor indices if the indices are a FakeTensor then don't require them to actually exist.
- Added TensorVariable.method_coalesce() implementation (unimplemented so it graph breaks - correct since the output tensor is dependent on the input data)
- merge_view_inputs() shouldn't be looking at storage of sparse tensors
- In insertable_tensor_check() don't treat sparse tensors as uniform.

And some drive-bys:
- Spelling nits
- Moved error text out of header file
- In has_metadata_mutation() only check StorageWeakRef if the result is going to be used.

Fixes #119780

This fixes about 71 of the 110 errors in #119780. The remaining issues hit follow-on errors (and should probably get new tasks once we refresh the error list)




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

laithsakka commented:
I wonder is shall i add sparce_coo_tensor to graph breaks for now in https://github.com/pytorch/pytorch/pull/120458

aorenste committed:
Update on "Fix dynamo issue "Failed running call_function <built-in method sparse_coo_tensor of type object at 0xDEADBEEF""

A small medley of fixes:
- When validating sparse tensor indices don't check numel() if it's symbolic.
- When validating sparse tensor indices if the indices are a FakeTensor then don't require them to actually exist.
- Added TensorVariable.method_coalesce() implementation (unimplemented so it graph breaks - correct since the output tensor is dependent on the input data)
- merge_view_inputs() shouldn't be looking at storage of sparse tensors
- In insertable_tensor_check() don't treat sparse tensors as uniform.

And some drive-bys:
- Spelling nits
- Moved error text out of header file
- In has_metadata_mutation() only check StorageWeakRef if the result is going to be used.

Fixes #119780

This fixes about 71 of the 110 errors in #119780. The remaining issues hit follow-on errors (and should probably get new tasks once we refresh the error list)




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]
